src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -56,22 +60,41 @@ internal class FindUsagesLSPContext : FindUsagesContext
         public override CancellationToken CancellationToken { get; }
 
         public FindUsagesLSPContext(
+            IProgress<object[]> progress,
             Document document,
             int position,
             IMetadataAsSourceFileService metadataAsSourceFileService,
             CancellationToken cancellationToken)
         {
+            _progress = progress;
             _document = document;
             _position = position;
             _metadataAsSourceFileService = metadataAsSourceFileService;
 
             CancellationToken = cancellationToken;
         }
 
-        public List<VSReferenceItem> GetReferences() => _resultsChunk;
+        public override async Task OnCompletedAsync()
+        {
+            using var _ = ArrayBuilder<VSReferenceItem>.GetInstance(out var referencesToReport);
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (_resultsChunk.IsEmpty())
+                {
+                    return;
+                }
+
+                referencesToReport.AddRange(_resultsChunk.ToArray());
```

Comment:
can we avoid this .ToArray()?  it doesn't seem necessary, and avoids allocations.